### ISSUE TITLE
Break up tasks into groups where tasks from one group may generate input for tasks in the next

### ISSUE THREAD

**Issue by renatoathaydes** [renatoathaydes/dartle#2]
This is a common problem:

* Task A generates output.
* Task B consumes Task A's output.

The problem is that Dartle currently computes all tasks inputs/outputs before running anything, hence it cannot know if a task needs to run when its inputs depend on the output of another task.

To fix this, we could break up the tasks into groups (automatically or manually?) so that Dartle computes which tasks need to run in stages.

For example, in the case mentioned above, tasks A and B are in separate groups, say group 1 contains Task A, group 2 contains Task B... now, to beging with, Dartle will only compute which tasks need to run on group 1... once it's done, Dartle computes which tasks need to run on group 2, and so on.

**Comment by renatoathaydes** [comment]
This will also enable `clean`-like tasks to run before anything else.

**Comment by renatoathaydes** [comment]
This was fixed by enforcing that when a task uses the outputs of another task as its inputs, an explicit dependency must exist between the tasks.

The problem of having a "group" of tasks run before any tasks in another group was fixed by introducing build phases.
